File: solutions/day1.py
def part1():
    with open("./data/1.txt") as f:
        data = f.read()
        sum = 0
        for line in data.split("\n"):
            first = None
            last = None
            for l in line:
                if l.isdigit():
                    if first is None:
                        first = l
                    last = l
            number = str(first) + str(last)
            sum += int(number)
        return print(sum)


def part2():
    numbers = {
        "one": 1,
        "two": 2,
        "three": 3,
        "four": 4,
        "five": 5,
        "six": 6,
        "seven": 7,
        "eight": 8,
        "nine": 9,
    }
    with open("./data/1.txt") as f:
        data = f.read()
        sum = 0
        for line in data.split("\n"):
            l = []
            for i in range(len(line)):
                if line[i].isdigit():
                    l.append(int(line[i]))
                for j in range(i + 1, len(line) + 1):
                    substring = line[i:j]
                    if substring in numbers:
                        first = numbers.get(substring)
                        l.append(first)
            number = str(l[0]) + str(l[-1])
            sum += int(number)
        return sum


part1()
part2()

File: solutions/day1.rs
use std::fs::File;
use std::io;
use std::io::BufRead;
use std::io::BufReader;

fn part1() -> io::Result<()> {
    let file = File::open("./data/1.txt").expect("Failed to open");
    let buf_reader = BufReader::new(file);
    let mut sum = 0;
    for line in buf_reader.lines() {
        let mut first = None;
        let mut last = None;
        if let Ok(ref l) = line {
            for ch in l.chars() {
                if ch.is_digit(10) {
                    if first.is_none() {
                        first = Some(ch);
                    }
                    last = Some(ch);
                }
            }
        }
        if let (Some(first_number), Some(last_number)) = (first, last) {
            let number_str = first_number.to_string() + &last_number.to_string();
            match number_str.parse::<i32>() {
                Ok(num) => sum += num,
                Err(_) => {
                    eprintln!("Failed to parse '{}' as integer", number_str);
                }
            }
        } else {
            eprintln!("First or last number not found");
        }
    }
    println!("{}", sum);
    Ok(())
}

fn main() {
    let _ = part1();
}
